nixos/modules/services/security/certmgr.nix

```diff
@@ -0,0 +1,210 @@
+          action = "reload";
+          authority = {
+            file.path = secret "ca";
+          };
+          certificate = {
+            path = secret domain;
+          };
+          private_key = {
+            owner = "root";
+            group = "root";
+            mode = "0600";
+            path = secret "''${domain}-key";
+          };
+          request = {
+            CN = domain;
+            hosts = [ "mail.''${domain}" "www.''${domain}" ];
+            key = {
+              algo = "rsa";
+              size = 2048;
+            };
+            names = {
+              O = "Example Organization";
+              C = "USA";
+            };
+          };
+        };
+      }
+      '';
+      type = with types; (attrsOf (submodule ({ ... }: {
+        options = {
+          service = mkOption {
+            type = nullOr str;
+            default = null;
+            description = "The service on which to perform &lt;action> after fetching.";
+          };
+
+          action = mkOption {
+            type = str;
+            default = "nop";
+            description = "The action to take after fetching.";
+          };
+
+          # These ought all to be specified according to certmgr spec def.
+          authority = mkOption {
+            type = attrs;
+            description = "certmgr spec authority object.";
+          };
+
+          certificate = mkOption {
+            type = nullOr attrs;
+            description = "certmgr spec certificate object.";
+          };
+
+          private_key = mkOption {
+            type = nullOr attrs;
+            description = "certmgr spec private_key object.";
+          };
+
+          request = mkOption {
+            type = nullOr attrs;
+            description = "certmgr spec request object.";
+          };
+        };
+    })));
+      description = ''
+        Certificate specs as described by:
+        <link xlink:href="https://github.com/cloudflare/certmgr#certificate-specs" />
+        These will be added to the Nix store, so they will be world readable.
+      '';
+    };
+
+    mkTargetDirs = mkOption {
+      default = false;
+      type = types.bool;
+      description = "Determines whether target directories (including parents) should be created for keys and certs in declarative specs.";
+    };
+
+    svcManager = mkOption {
+      default = "systemd";
+      type = types.enum [ "circus" "command" "dummy" "openrc" "systemd" "sysv" ];
+      description = ''
+        The behavior to use when executing actions after fetching certs
+        See: <link xlink:href="https://github.com/cloudflare/certmgr#certmgryaml" />.
+      '';
+    };
+
+  };
+
+  config = {
+    assertions = with lib; singleton {
+      assertion = !any (hasAttrByPath [ "authority" "auth_key" ]) (mapAttrsToList (_: value: value) cfg.declSpecs);
+      message = "services.certmgr.declSpecs are added to the Nix store rendering them world readable. Use impSpecs if you want to include auth_key";
```

Comment:
Not entirely sure what you are saying here? Changed the assertion message slightly here: https://github.com/NixOS/nixpkgs/pull/44556/commits/e77155cbf3eafde44bc923e6f19b0e974a86c8f2 .. The point is that setting `auth_key` would certainly place a secret in nix-store, which we want to prevent. `auth_key_file` can be safely used, if set to a non-store-path.